core/src/avm2/specification.rs

```diff
@@ -0,0 +1,525 @@
+    #[serde(rename = "prototype")]
+    prototype: Option<TraitList>,
+}
+
+#[derive(Default)]
+struct ClassStubs {
+    methods: FnvHashSet<Cow<'static, str>>,
+    getters: FnvHashSet<Cow<'static, str>>,
+    setters: FnvHashSet<Cow<'static, str>>,
+}
+
+impl ClassStubs {
+    pub fn for_class(class_name: &str, stubs: &FnvHashSet<&Stub>) -> Self {
+        let mut result = ClassStubs::default();
+
+        for stub in stubs
+            .iter()
+            .filter(|s| s.avm2_class() == Some(Cow::Borrowed(class_name)))
+        {
+            match stub {
+                Stub::Avm2Method { method, .. } => {
+                    result.methods.insert(method.clone());
+                }
+                Stub::Avm2Getter { property, .. } => {
+                    result.getters.insert(property.clone());
+                }
+                Stub::Avm2Setter { property, .. } => {
+                    result.setters.insert(property.clone());
+                }
+                _ => {}
+            }
+        }
+
+        result
+    }
+
+    pub fn has_method(&self, name: &str) -> bool {
+        self.methods.contains(name)
+    }
+
+    pub fn has_getter(&self, name: &str) -> bool {
+        self.getters.contains(name)
+    }
+
+    pub fn has_setter(&self, name: &str) -> bool {
+        self.setters.contains(name)
+    }
+}
+
+impl Definition {
+    fn from_class<'gc>(
+        class_object: ClassObject<'gc>,
+        activation: &mut Activation<'_, 'gc>,
+        stubs: &ClassStubs,
+    ) -> Self {
+        let mut definition = Self::default();
+        let class = class_object.inner_class_definition();
+
+        if class.read().is_final() {
+            definition
+                .classinfo
+                .get_or_insert_with(Default::default)
+                .is_final = true;
+        }
+        if !class.read().is_sealed() {
+            definition
+                .classinfo
+                .get_or_insert_with(Default::default)
+                .dynamic = true;
+        }
+        if let Some(super_name) = class
+            .read()
+            .super_class_name()
+            .as_ref()
+            .and_then(|n| n.local_name())
+        {
+            if &super_name != b"Object" {
+                definition
+                    .classinfo
+                    .get_or_insert_with(Default::default)
+                    .extends = Some(super_name.to_string());
+            }
+        }
+
+        let prototype = class_object.prototype();
+        let prototype_base = prototype.base();
+        let prototype_values = prototype_base.values();
+        for (key, value) in prototype_values.as_hashmap().iter() {
+            let name = match key {
+                DynamicKey::String(name) => *name,
+                DynamicKey::Uint(key) => {
+                    AvmString::new_utf8(activation.context.gc_context, key.to_string())
+                }
+                DynamicKey::Object(object) => {
```

Comment:
I saw this while using this code yesterday and it didn't really make sense to me. I don't think this can happen.

Reply:
Probably not, but there's no real harm in having it here.